### PR TITLE
Rogue Legacy: World folder clean up and generation improvements.

### PR DESCRIPTION
**Pull request by ThePhar**
Completely refactored the logic for Rogue Legacy. It is backwards compatible with the latest release, but fixes a few bugs with early vendors not working and also is easier to read. Mostly as a foundation for the next Rogue Legacy release so I can work with something not a pain to debug/read.

## What is this fixing or adding?
- Renamed all Leagcy*X* to RL*X*. (e.g. RLWorld, RLWebWorld, RLItem, etc.)
- Change `early` vendors to be in the first sphere, forced local.
- Re-introduce `early` architect setting, to work like `early` vendors.
- Reclassify item classifications for better clarity in logic.
- Increment `data_version` to 4, as some locations have been renamed for better clarity.
- Fairy chests per zone can now be set to 0. Min chests per zone was increased to 20 to compensate.
- Fairy chests now logically require an Enchantress + Movement rune OR Dragons for increased odds of completing challenges.
- Changed `AdditionalNames` to be an `OptionSet` instead of `OptionList`.
- Add locations into regions per zone and add rules to regions to reduce number of lambdas needed for logic.
- Added bug report link to `RLWebWorld`.
- Tweaked documentation to include more up to date information about randomizer.

## How was this tested?
Has been tested on latest version of Rogue Legacy Randomizer v0.9.2 and has successful beta tests completed in AP discord.

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

**Comment by ThePhar** [comment]
Creating PR for running tests and also awaiting completing of beta test before moving out of Draft.

File: worlds/rogue_legacy/docs/rogue-legacy_en.md [hunk 1]
```diff
@@ -28,25 +28,8 @@ provides an alternative one to the default values.
...
-
-- Content/
-
-Then copy the contents of the CustomContent directory in your Rogue Legacy Randomizer into the newly copied Content
-directory and overwrite all files.
-
-**BE SURE YOU ARE REPLACING THE COPIED FILES IN YOUR ROGUE LEGACY RANDOMIZER DIRECTORY AND NOT REPLACING YOUR ROGUE
-LEGACY FILES!**
+Please read the README file on the 
+[Rogue Legacy Randomizer GitHub](https://github.com/ThePhar/RogueLegacyRandomizer/blob/master/README.md) page for upto
```
Comment (Berserker66) [inline worlds/rogue_legacy/docs/rogue-legacy_en.md:34]: "upto" typo?

**Review by black-sliver — COMMENTED** [review]
You remove some locations and items from datapackage. Unless they are guaranteed unused, this breaks playing/finishing (hinting) old seeds on the new host, e.g.
```
        "Castle Hamson - Chest 51": 91650,
        ...
        "Castle Hamson - Chest 100": 91699,
        "Castle Hamson - Fairy Chest 16": 91415,
        ...
        "Castle Hamson - Fairy Chest 50": 91449,
```
and a lot more are gone from the new datapackage.
Were they previously unused?

File: worlds/rogue_legacy/__init__.py [hunk 2]
```diff
@@ -1,176 +1,270 @@
...
+        return getattr(self.world, name)[self.player]
 
     def fill_slot_data(self) -> dict:
-        slot_data = self._get_slot_data()
-        for option_name in legacy_options:
-            option = getattr(self.world, option_name)[self.player]
+        slot_data = {}
+        for option_name in rl_options:
+            option = self.setting(option_name)
             slot_data[option_name] = option.value
```
Comment (black-sliver) [inline worlds/rogue_legacy/__init__.py:53]: I would probably write it like this:
```suggestion
        slot_data = {option_name: self.setting(option_name).value for option_name in rl_options}
```

File: worlds/rogue_legacy/__init__.py [hunk 3]
```diff
@@ -1,176 +1,270 @@
...
+        return RLItem(name, data.classification, data.code, self.player)
 
     def set_rules(self):
         set_rules(self.world, self.player)
+
+    def create_regions(self):
+        create_regions(self.world, self.player)
+        self.__place_events()
+
+    def __place_events(self):
```
Comment (black-sliver) [inline worlds/rogue_legacy/__init__.py:235]: any reason why this is `__place_events` and not `_place_events` ? Python does special handling for `__`, which makes single `_` equivalent to "protected", double `__` equivalent to private.

**Review by black-sliver — COMMENTED** [review]
Two more cleanup things

File: worlds/rogue_legacy/__init__.py (hunk 3 above)
Comment (ThePhar) [reply worlds/rogue_legacy/__init__.py:235]: Because it's supposed to be only used by this class, hence private.

**Comment by ThePhar** [comment]
> You remove some locations and items from datapackage. Unless they are guaranteed unused, this breaks playing/finishing (hinting) old seeds on the new host, e.g.
> 
> ```
>         "Castle Hamson - Chest 51": 91650,
>         ...
>         "Castle Hamson - Chest 100": 91699,
>         "Castle Hamson - Fairy Chest 16": 91415,
>         ...
>         "Castle Hamson - Fairy Chest 50": 91449,
> ```
> 
> and a lot more are gone from the new datapackage. Were they previously unused?

Yes. These locations were generated in the table, but never actually created in game. The option was limited to up to 30 chests per zone, and up to fairy 15 chests per zone. I just removed those not used locations.

File: worlds/rogue_legacy/__init__.py (hunk 3 above)
Comment (ThePhar) [reply worlds/rogue_legacy/__init__.py:235]: I don't see any world where something should inherit from this class.

File: worlds/rogue_legacy/docs/rogue-legacy_en.md (hunk 1 above)
Comment (ThePhar) [reply worlds/rogue_legacy/docs/rogue-legacy_en.md]: English is hard. Fixed.

File: worlds/rogue_legacy/__init__.py (hunk 3 above)
Comment (black-sliver) [reply worlds/rogue_legacy/__init__.py:235]: Oh, maybe let me rephrase. Python will rename the function from `__place_events` to `_RLWorld__place_events` to ensure it being "private". We normally only do that in python if we explicitly want something to be private for a reason. So if I see `__` I'm wondering what the reason for that is, and not the other way around.

Comment (ThePhar) [reply worlds/rogue_legacy/__init__.py:235]: I understand the implementation details, I'm talking methodology. In my experience in other projects and languages, you typically want an explicit reason for making something more accessible, not less. You want to reduce the amount of API that isn't needed and make it easier if you need to refactor later.

Protected would should be fine, but I don't see a reason for it to be protected vs private if `RLWorld` isn't to be inherited from.

Might be different in philosophy in how we determine access modifiers.

File: worlds/rogue_legacy/__init__.py (hunk 2 above)
Comment (ThePhar) [reply worlds/rogue_legacy/__init__.py]: Made some tweaks to this to simplifies it more. Thanks!

**Review by black-sliver — APPROVED** [review]
